Move the remaining controllers to shared informers

smarterclayton opened the pull request:
Completes work done in 1.6 to move the last two hold outs to shared informers - tokens controller and scheduler. Adds a few more tools to allow informer reuse (like filtering the informer, or maintaining a mutation cache).

The mutation cache is identical to #45838 and will be removed when that merges

@ncdc @deads2k extracted from openshift/origin#14086

smarterclayton committed:
Add a filtering resource handler for informers

Allows an informer consumer to easily filter a set of changes out,
possibly to maintain a smaller cache or to only operate on a known set
of objects.

smarterclayton committed:
Switch the tokens controller to use shared informers

Tokens controller previously needed a bit of extra help in order to be
safe for concurrent use. The new MutationCache allows it to keep a local
cache and still use a shared informer. The filtering event handler lets
it only see changes to secrets it cares about.

smarterclayton committed:
Scheduler should use shared informer for pods

Previously, the scheduler created two separate list watchers. This
changes the scheduler to be able to leverage a shared informer, whether
passed in externally or spawned using the new in place method. This
removes the last use of a "special" informer in the codebase.

Allows someone wrapping the scheduler to use a shared informer if they
have more information avaliable.

smarterclayton committed:
generated: bazel

deads2k commented on pkg/controller/serviceaccount/tokens_controller.go:
```diff
@@ -71,7 +68,7 @@ type TokensControllerOptions struct {
 
 // NewTokensController returns a new *TokensController.
-func NewTokensController(cl clientset.Interface, options TokensControllerOptions) *TokensController {
+func NewTokensController(serviceAccounts, secrets cache.SharedIndexInformer, cl clientset.Interface, options TokensControllerOptions) *TokensController {
```
Type the informers please.  We have generated types for SecretInformer and ServiceAccoutnInfomer

deads2k commented on pkg/controller/serviceaccount/tokens_controller.go:
```diff
@@ -91,44 +88,39 @@ func NewTokensController(cl clientset.Interface, options TokensControllerOptions
+	secretCache := secrets.GetIndexer()
+	e.updatedSecrets = controller.NewIntegerResourceVersionMutationCache(secretCache, secretCache)
+	e.secretController = secrets.GetController()
+	secrets.AddIndexers(cache.Indexers{"namespace": cache.MetaNamespaceIndexFunc})
```
this indexer is already present in the generated informer

smarterclayton replied:
are we guaranteed to get this index in every possible SecretInformer?

deads2k commented on pkg/controller/serviceaccount/tokens_controller.go:
```diff
@@ -91,44 +88,39 @@ func NewTokensController(cl clientset.Interface, options TokensControllerOptions
-				return e.client.Core().Secrets(metav1.NamespaceAll).List(options)
+	secretCache := secrets.GetIndexer()
+	e.updatedSecrets = controller.NewIntegerResourceVersionMutationCache(secretCache, secretCache)
+	e.secretController = secrets.GetController()
```
I doubt you really need a controller now.  Probably just hasSynced

deads2k commented on pkg/controller/serviceaccount/tokens_controller.go:
```diff
@@ -91,44 +88,39 @@ func NewTokensController(cl clientset.Interface, options TokensControllerOptions
-		},
-		&v1.ServiceAccount{},
-		options.ServiceAccountResync,
+	e.serviceAccounts = serviceAccounts.GetStore()
```
service accounts should be able to use a Lister

deads2k commented on pkg/controller/serviceaccount/tokens_controller.go:
```diff
@@ -142,7 +134,7 @@ type TokensController struct {
 
 	serviceAccounts cache.Store
-	secrets         cache.Indexer
+	updatedSecrets  controller.MutationCache
```
godoc this.  mutation caches aren't common

deads2k commented on pkg/controller/serviceaccount/tokens_controller.go:
```diff
@@ -166,29 +158,23 @@ type TokensController struct {
-	for !e.serviceAccountController.HasSynced() || !e.secretController.HasSynced() {
-		time.Sleep(100 * time.Millisecond)
+	// Wait for the stores to fill
+	if !cache.WaitForCacheSync(stopCh, e.serviceAccountController.HasSynced, e.secretController.HasSynced) {
```
there's a different method upstream that should drop a log message for you.  `pkg/contrrollers` maybe?

deads2k commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -136,17 +137,58 @@ func NewConfigFactory(
-			UpdateFunc: c.updatePodInCache,
-			DeleteFunc: c.deletePodFromCache,
+	scheduledPodIndexer := podInformer.Informer().GetIndexer()
+	scheduledPodIndexer.AddIndexers(cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
```
Please call on the informer.  This isn't always possible and the informers panic for you instead of breaking.

deads2k commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -136,17 +137,58 @@ func NewConfigFactory(
-			DeleteFunc: c.deletePodFromCache,
+	scheduledPodIndexer := podInformer.Informer().GetIndexer()
+	scheduledPodIndexer.AddIndexers(cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
+	c.scheduledPodPopulator = podInformer.Informer().GetController()
```
this is suspect.  Some controller calls don't work right on informers anymore.  I remember wiring them to dummy.

deads2k commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -500,19 +539,47 @@ func getNodeConditionPredicate() corelisters.NodeConditionPredicate {
-func (factory *ConfigFactory) createUnassignedNonTerminatedPodLW() *cache.ListWatch {
-	selector := fields.ParseSelectorOrDie("spec.nodeName==" + "" + ",status.phase!=" + string(v1.PodSucceeded) + ",status.phase!=" + string(v1.PodFailed))
-	return cache.NewListWatchFromClient(factory.client.Core().RESTClient(), "pods", metav1.NamespaceAll, selector)
+// unassignedNonTerminatedPod selects pods that are unassigned and non-terminal.
```
marker for how far I got

ncdc commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -136,19 +137,59 @@ func NewConfigFactory(
-			AddFunc:    c.addPodToCache,
-			UpdateFunc: c.updatePodInCache,
-			DeleteFunc: c.deletePodFromCache,
+	podInformer.Informer().AddIndexers(cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
```
The generated shared informers have the namespace index enabled by default - see e.g. https://github.com/kubernetes/kubernetes/blob/a86fabb9d22b94f61b967510f4fb82dd98946bd5/pkg/client/informers/informers_generated/externalversions/core/v1/pod.go#L56

ncdc replied:
Oh, I see this isn't one of the generated ones now that I've read more. Can you add a comment?

ncdc commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -136,19 +137,59 @@ func NewConfigFactory(
+			},
+		},
+	)
+	c.scheduledPodLister = corelisters.NewPodLister(podInformer.Informer().GetIndexer())
```
```go
c.scheduledPodLister = podInformer.Lister()
```

ncdc replied:
Nevermind, can you add a comment why you're doing this?

ncdc commented on plugin/pkg/scheduler/factory/factory_test.go:
```diff
@@ -55,6 +55,7 @@ func TestCreate(t *testing.T) {
 		v1.DefaultSchedulerName,
 		client,
 		informerFactory.Core().V1().Nodes(),
+		informerFactory.Core().V1().Pods(),
```
Shouldn't this be your custom podInformer defined in factory.go?

smarterclayton replied:
Test verifies that both work

deads2k commented:
lgtm.  The mutation cache should be in master now.

smarterclayton commented:
Testing to see what the problem with service account is

smarterclayton committed:
Mutation cache should support retrieving items from ByIndex()

Allows tokens controller to observe updates

smarterclayton committed:
Move pkg/util/cache to apimachinery

Will be used by client-go as well

smarterclayton committed:
Change LRUExpireCache to use hashicorp cache to expose Keys()

Removes the spawning of goroutines in the cache (which could be a
hotspot for anything in the critical path) as well.

smarterclayton committed:
Update consumers of LRUExpireCache

deads2k commented:
lru updates look good.

deads2k commented:
can you squash up the LRU changes into one commit?

deads2k commented on staging/src/k8s.io/client-go/tools/cache/mutation_cache.go:
```diff
@@ -128,6 +140,32 @@ func (c *mutationCache) ByIndex(name string, indexKey string) ([]interface{}, er
+		fn := c.indexer.GetIndexers()[name]
+		// Keys() is returned oldest to newest, so full traversal does not alter the LRU behavior
+		for _, key := range c.mutationCache.Keys() {
+			if keySet.Has(key.(string)) {
```
Skipping this get changes LRU position.   I'm ok with it, but its worth noting.

smarterclayton replied:
urg.  hrm.  erk.

deads2k commented on staging/src/k8s.io/client-go/tools/cache/mutation_cache.go:
```diff
@@ -128,6 +140,32 @@ func (c *mutationCache) ByIndex(name string, indexKey string) ([]interface{}, er
+			}
+			elements, err := fn(updated)
+			if err != nil {
+				return nil, err
```
I would log and continue.  I think this is a nice to have, not a "fail me" layer.

deads2k commented:
I'd like the failure ignored, then lgtm.

smarterclayton committed:
Move client-go/util/clock to apimachinery/pkg/util/clock

For reuse

smarterclayton committed:
Refactor move of client-go/util/clock to apimachinery

smarterclayton commented:
Tagging as per comments and green.

wojtek-t commented:
@smarterclayton @deads2k @ncdc - I reverted the commit about scheduler in https://github.com/kubernetes/kubernetes/pull/46199
You can take a look into https://github.com/kubernetes/kubernetes/issues/46196 for more details.
